README.md
# Factory Health Monitor - Authentication, State Management, and Persistence

## Submission Details and Changes:

- Frontend Changes [here](./native-app/changes.md) 
- Backend Changes [here](./backend/changes.md)


## Overview

You are provided with a starter project that includes a React Native mobile app and a backend API, each in their respective folders. Each folder contains a README with instructions on how to run the builds. The goal of this coding challenge is to add authentication and session management to the app, improve the state management of the data returned by the API, and implement a persistence layer on the backend. When a user logs in, their history of data points and scores should be retrieved and displayed.

### The Application

The application you'll be working on is a tool for evaluating the health of various machines in an automobile manufacturing plant. This plant features a range of machines, each with specific data points that you will use to assess the condition of the production process.

#### Machines and Their Variables

1. **Welding Robots**

   - Welding Robot Error Rate
   - Welding Arm Vibration Level
   - Electrode Wear
   - Gas Shielding Pressure
   - Welding Wire Feed Rate
   - Arc Stability
   - Weld Seam Width
   - Cooling System Efficiency

2. **Painting Stations**

   - Paint Flow Rate
   - Paint Pressure
   - Paint Color Consistency
   - Paint Nozzle Condition

3. **Assembly Lines**

   - Part Alignment Accuracy
   - Assembly Line Speed
   - Component Fitting Tolerance
   - Conveyor Belt Speed

4. **Quality Control Stations**
   - Inspection Camera Calibration
   - Inspection Light Intensity
   - Inspection Software Version
   - Inspection Criteria Settings

### Time Limit

You are expected to spend approximately 3 hours on this challenge. While we understand that building a app can be a never-ending task, we're interested in seeing how far you can get and how you prioritize tasks to maximize quality/delivery within the given time frame.

## Repository Structure

The repository is structured as follows:

```
├── native-app/
│   ├── source code files...
│   ├── README.md               # The README file for running the React Native Mobile app
│   └── ...
│
├── backend/
│   ├── source code files...
│   ├── README.md               # The README file for running the API Backend
│   └── ...
│
├── MachineHealth.apk           # The compiled android app for running on a device or emulator (if helpful)
├── README.md                   # This README file
```

## Getting Started

1. Fork this repository to your GitHub account.
2. Clone your forked repository to your local machine.

## Running the App Locally

To run the Machine Health Evaluation app locally, you'll need to set up and run both the React Native app and the API backend separately. Each of the respective folders are in this director and each have their own README files to help you get started.

To use the Machine Health Evaluation app locally, you'll need to keep both the API and the React Native app running simultaneously. It's recommended to open separate terminal/command windows for each and run them in parallel.

- In one terminal window, navigate to the `backend` folder and run the API backend.
- In another terminal window, navigate to the `native-app` folder and run the React Native app.

Keep in mind that the React Native app relies on the API to fetch and calculate machine health data. Ensure that the API is accessible to the app for it to function correctly.

For specific details on running the API and React Native app, refer to their respective README files in their respective folders.

## Requirements

1. **Authentication and Session Management:**

   - Implement user authentication in the Expo mobile app. Users should be able to log in securely using credentials.
   - Manage user sessions and ensure that users remain authenticated between app sessions.

2. **Data State Management:**

   - Review and update (as needed) the existing state management for data returned by the API. Feel free to bring in a state management library if it'd be helpful.
   - Ensure that the app efficiently manages and updates the state when new data points and scores are fetched.

3. **Persistence Layer on the Backend:**

   - Implement a persistence layer on the backend to store historical data points and scores for each user/machine.
   - When a new data point or score is recorded, ensure it is stored in the persistence layer for future retrieval.

4. **Stretch Goals (Optional):**
   - Implement a section to show the history of scores with trends over time.
   - Include visualizations such as charts or graphs to represent the trends in machine health scores.

## Guidelines

- Use the provided backend API and Expo mobile app as a starting point.
- Follow best practices for authentication and session management in React Native.
- Enhance the state management to ensure a clean and organized flow of data.
- Implement a persistence layer on the backend for storing historical data points and scores.
- Document any changes made to the existing codebase and provide a brief explanation of your architectural decisions.

## Submission

- Provide a link to your Git repository containing the updated project.
- Include instructions on how to run the modified builds locally.

## Evaluation Criteria

- Implementation of Authentication and Session Management
- Improvement of Data State Management
- Implementation of a Persistence Layer on the Backend
- Adherence to Best Practices
- Documentation
- Overall Code Quality

**Note:**
This project is designed to take approximately 3 hours, but there is no strict time limit. We are interested in seeing how far you can get and the quality of your final deliverable. Focus on completing the core requirements before attempting the stretch goals. Feel free to make reasonable assumptions if certain details are not explicitly provided. If you encounter challenges, document them and describe how you would address them with more time.


backend/changes.md
## Backend Changes
----

## Pre-Requisites

- For persistence we used MongoDB. Follow any of these three to setup your mongo instance. 
    - Mongo on cloud using [Atlas UI](https://www.mongodb.com/docs/atlas/getting-started/)
    - Docker instance, [setup instructions](#running-mongodb-in-docker) [Preferred]

## Set Mongo URI

- Set `MONGODB_URI` which is used for persistence in [`db.ts`](./netlify/functions/api/db.ts); 

## Authentication Middleware:

JSON Web Token (jsonwebtoken) is used for authentication middleware. The setup is done in [`authMiddleware.ts`](../backend/netlify/functions/api/authMiddleware.ts). You can generate your own token to update line `15` or continue using this.


## Installation

Follow these steps to set up:

1. Navigate to the project directory:

   ```bash
   cd api
   ```
2. Install dependencies using Yarn (or npm if you prefer):

   ```bash
   yarn
   ```
### Starting the API
To start the API, run the following command:
```bash
yarn start
```
## Modifications

### Authentication and Session Management:

MongoDB is used here; a collection called `users` will store username and hashed password as a user registers.
More on [this](../backend/netlify/functions/api/user.ts).

#### User Registration

Register a user by sending a POST request to the `/register` endpoint. Here's an example using cURL:

```bash
curl -X POST -H "Content-Type: application/json" -d '{
  {
  "username": "testUser",
  "password": "testPassword"
}' http://localhost:3001/register
```

Sample success response:
```Json
{
    "message": "User Created Successfully!"
}
```

#### User Login
Login in app by sending a POST request to the `/login` endpoint. Here's an example using cURL:

```bash
curl -X POST -H "Content-Type: application/json" -d '{
  {
  "username": "testUser",
  "password": "testPassword"
}' http://localhost:3001/login
```
The response will provide an authentication token:

```Json
{
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InJvbGUyIiwiaWF0IjoxNzAzMDM1Mzc1fQ.DMn87z-On9FDKg1TMXkkFRgxOVmLhZFizWc8_T3y5h8"
}
```
### Persistence Layer on the Backend:

We use two collections in mongo for the whole persistence logic: `users` & `datapoints`.

- `users` is to store the user records
- `datapoints` is to store the historical data records.

Screenshot of data in mongodb:
![users](./assets/users.png)
![dataPoints](./assets/datapoints.png)


### Stretch Goals (history of scores):
It includes a feature to visualize trends in machine health scores.

1. The GetAPI endpoint `historyDataPoints` is implemented in [`app.ts`](./app.ts)
2. Data Manipulation for Visualization:
  Data manipulation for visualization of trends in machine health scores is done in [`syncDataPoints.ts`](../backend/syncDataPoints.ts).
3. MongoDB Storage: Manipulated data is saved into MongoDB for historical tracking. The file responsible for this operation is [`syncDataPoints.ts`](../backend/syncDataPoints.ts).

The response includes data for different machines, each with corresponding dates and scores which is used by frontend to make visualizations.
``` Json
[
    {
        "label": "assemblyLine",
        "labels": [
            "2023-12-19",
            "2023-12-18",
            "2023-12-9",
            "2023-12-19",
            "2023-12-19",
            "2023-12-19"
        ],
        "data": [
            "0.00",
            "72.22",
            "72.22",
            "72.22",
            "72.22",
            "72.22"
        ]
    },
    {
        "label": "qualityControlStation",
        "labels": [
            "2023-12-19"
        ],
        "data": [
            "20.00"
        ]
    },
    {
        "label": "weldingRobot",
        "labels": [
            "2023-12-23",
            "2023-12-1",
            "2023-12-19",
            "2023-12-19",
            "2023-12-19"
        ],
        "data": [
            "88.19",
            "88.19",
            "88.19",
            "88.19",
            "88.19"
        ]
    }
]
```

## Running MongoDB in Docker

To run mongo in docker; 
- [Installation guideline for Mac](https://docs.docker.com/desktop/install/mac-install/) - Much straightforward, just install `Docker Desktop` 
- [Installation guideline for Linux](https://docs.docker.com/desktop/install/linux-install/)
- [Installation guideline for Window](https://docs.docker.com/desktop/install/windows-install/)

If you are using `Docker Desktop` you should automatically have `docker-compose` utility installed. If not refer this [guide](https://docs.docker.com/compose/install/).

- Mongo Client [Optional]

To view mongo, I had included a web based interface with docker called `mongo-express`. If no you can use an offline client such as [`mongo compass`](https://www.mongodb.com/try/download/compass). Incase you are using Atlas - cloud instance, it should be already giving you an user interface to play around :).

Once you have this installation done -

- `cd EngineeringChallenge`
- Run docker compose -
`docker-compose -f docker-compose.yml up -d`

* [`docker-compose.yml`](../docker-compose.yml) defines `root` as mongo username and `password` as mongo password. Your final connection string would look like `mongodb://root:password@localhost:27017/`

* Visit [localhost:8088](http://localhost:8088/) for mongo-express web interface.

To stop mongodb:
- `docker-compose -f docker-compose.yml down`



backend/syncDataPoints.ts
import { historyDataModel } from '../backend/netlify/functions/api/historyData';
import { format } from 'date-fns';

export const syncToCloud = (requestBody: any, result: any) => {
    const machinesData = requestBody.machines;
    const scoresData = result.machineScores;

    // Mapping machines data
    const machineIds = Object.keys(machinesData);
    const machinesHistoryData = machineIds.map((machineId) => {
        const scoreValue = scoresData[machineId] ;
        const today = format(new Date(), 'yyyy-MM-dd');

        return new historyDataModel({
            machineId: machineId,
            userId: 'role2',
            timestamp: today,  
            score: scoreValue,
            dataPoints: {
                [machineId]: machinesData[machineId]
            }
        });
    });



    return machinesHistoryData;
};


backend/netlify/functions/api/userService.ts
import { UserModel, IUser } from './user';
import bcrypt from 'bcrypt';

interface UserError {
  type: string;
  message: string;
  error?: any;
}

/**
 * Registers a new user with the provided username and password.
 * @param {string} username - The username for the new user.
 * @param {string} password - The password for the new user.
 * @example
 * // Example usage:
 * const registrationResult = await registerUser("testUser", "testUserPassword");
 */
export async function registerUser(
    username: string,
    password: string
  ): Promise<{ error: UserError | null; user: IUser | null }> {
    const maxRetries = 3;
    const retryDelay = 5000; // 5 seconds delay between retries
  
    let retryCount = 0;
  
    while (retryCount < maxRetries) {
      try {
        const hash = await bcrypt.hash(password, 10);
  
        const user = new UserModel({ username, password:hash});
        const savedUser = await user.save();
  
        return { error: null, user: savedUser as IUser }; // Explicitly cast to IUser
      } catch (error: any) {
        // Handle unique constraint violation (duplicate key error)
        if (error.code === 11000 || error.code === 11001) {
          return { error: { type: 'duplicate_username', message: 'Username already exists' }, user: null };
        }
  
        retryCount++;
        // Add a longer delay before retrying
        await new Promise(resolve => setTimeout(resolve, retryDelay));
      }
    }
  
    return { error: { type: 'save_error', message: 'Error saving user - max retries reached' }, user: null };
  }
  

  export async function loginUser(
    username: string,
    password: string,
    callback: (error: UserError | null, user: IUser | null) => void
  ): Promise<void> {
    try {
      const user: IUser | null = await UserModel.findOne({ username });
  
      if (!user) {
        return callback({ type: 'user_not_found', message: 'User not found' }, null);
      }
  
      bcrypt.compare(password, user.password, (compareError, isMatch) => {
        if (compareError) {
          return callback({ type: 'compare_error', message: 'Error comparing passwords', error: compareError }, null);
        }
  
        if (!isMatch) {
          return callback({ type: 'incorrect_password', message: 'Incorrect password' }, null);
        }
  
        callback(null, user);
      });
    } catch (error) {
      callback({ type: 'find_error', message: 'Error finding user', error }, null);
    }
  }
  


backend/netlify/functions/api/authMiddleware.ts
//Middleware for authenticating a user using a JWT token.

import { Request, Response, NextFunction } from 'express';
import jwt from 'jsonwebtoken';

// Extend Request type to include a user property
interface AuthenticatedRequest extends Request {
    user?: any; 
  }
  
export function authenticateToken(req: AuthenticatedRequest, res: Response, next: NextFunction) {
  const token = req.header('Authorization');
  if (!token) return res.status(401).send('Access denied');

  jwt.verify(token, 'e0NldHbb/VaWwH7B9J0kPM6B5HSPfKHOkeo4KF5VeQji8neWwTk7oI7kp44awOdPzvHaToJT9hJ36NkNA7VPZ1LIj/7TYLwloPUVTp/VhsAky9874XeQoaZOUUT2I+tOL3L0z8BnEbN2+uyH6kvKkNMUHLKWsRx1OxKf/oXaQt5X9P56lF7rPPuoY3JO97xgFvLaEY6393B1wnEiVWqHPW9xeMz/XOBIDiGQvHTAopAlg61UKJmHRaA6wfAh4HGQldMojIHPAat5zyTUAsjnkGNSV9BfCbCqKiOxbXE3rRC/4R5+2edpMHMLqAjq55B2S636m+Q0kTeRuiSNN+mH2Q==', (err: any, user: any) => {
    if (err) return res.status(403).send('Invalid token');
    req.user = user;
    next();
  });
}


backend/app.ts
import express, { Request, Response } from 'express';
import { getMachineHealth } from './machineHealth';
import { registerUser, loginUser } from '../backend/netlify/functions/api/userService';
import jwt from 'jsonwebtoken';
import cors from 'cors'; 
import mongoDb from './netlify/functions/api/db'
import { historyDataModel} from '../backend/netlify/functions/api/historyData'; 
import {syncToCloud} from './syncDataPoints'


const app = express();
const port = 3001;

// Middleware to parse JSON request bodies
app.use(express.json());
// Enable CORS for all routes
app.use(cors());

// Endpoint to get machine health score
app.post('/machine-health', async (req: Request, res: Response) => {
  const result = getMachineHealth(req);
  if (result.error) {
    res.status(400).json(result);
  } else {
    res.json(result);
    //push to mongo
    let historyDATA= syncToCloud(req.body,result)
    try {
     for (const data of historyDATA) {
        const savedData = await data.save();
    }
    } catch (error) {
      console.log('Error saving historical data:', error);
    }
  }
});

//Fetching  dataPoints stored in mongoDB
app.get('/historyDataPoints', async (req, res) => {
  try {
    let finalData=[]
    const fetchedData = await historyDataModel.distinct('machineId');
    for (const machineId of fetchedData) {
      // Find documents for each machineId
      const documents = await historyDataModel.find({ machineId });
      // Extract timestamps and scores
      const timestamps = documents.map(doc => doc.timestamp);
      const scores = documents.map(doc => doc.score);
      finalData.push({'label':machineId,'labels':timestamps,'data':scores})
    }
    res.json(finalData);
  } catch (error) {
    console.error('Error fetching data from MongoDB:', error);
    res.status(500).json({ error: 'Internal Server Error' });
  }
});

app.get('/hello', (req, res) => res.send("Hello World"));

app.post('/register', async (req: Request, res: Response) => {
  const { username, password } = req.body;

  try {
    const { error, user } = await registerUser(username, password);
    if (error) {
      return res.status(400).json({ error: error.message });
    }
    res.json({message: 'User Created Successfully!' });
  } catch (error) {
    res.status(500).json({ error: 'Internal Server Error' });
  }
});

app.post('/login', (req: Request, res: Response) => {
  const { username, password } = req.body;

  loginUser(username, password, (error, user) => {
    if (error) {
      return res.status(500).json({ error: 'Internal Server Error' });
    }

    if (!user) {
      return res.status(401).json({ error: 'Invalid username or password' });
    }

    const token = jwt.sign({ username: user.username }, 'e0NldHbb/VaWwH7B9J0kPM6B5HSPfKHOkeo4KF5VeQji8neWwTk7oI7kp44awOdPzvHaToJT9hJ36NkNA7VPZ1LIj/7TYLwloPUVTp/VhsAky9874XeQoaZOUUT2I+tOL3L0z8BnEbN2+uyH6kvKkNMUHLKWsRx1OxKf/oXaQt5X9P56lF7rPPuoY3JO97xgFvLaEY6393B1wnEiVWqHPW9xeMz/XOBIDiGQvHTAopAlg61UKJmHRaA6wfAh4HGQldMojIHPAat5zyTUAsjnkGNSV9BfCbCqKiOxbXE3rRC/4R5+2edpMHMLqAjq55B2S636m+Q0kTeRuiSNN+mH2Q==');

    res.json({ token });
  });
});


app.listen(port, () => {
  console.log(`API is listening at http://localhost:${port}`);
});

mongoDb.on('error', (err) => {
  console.error('MongoDB connection error:', err);
});

mongoDb.once('open', () => {
  console.log('Connected to MongoDB');
});
